compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -220,19 +221,53 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         };
 
         match (&mut val, field.abi) {
-            (OperandValue::Immediate(llval), _) => {
+            (
+                OperandValue::Immediate(llval),
+                Abi::Scalar(_) | Abi::ScalarPair(..) | Abi::Vector { .. },
+            ) => {
                 // Bools in union fields needs to be truncated.
                 *llval = bx.to_immediate(*llval, field);
                 // HACK(eddyb) have to bitcast pointers until LLVM removes pointee types.
-                *llval = bx.bitcast(*llval, bx.cx().immediate_backend_type(field));
+                let ty = bx.cx().immediate_backend_type(field);
+                if bx.type_kind(ty) == TypeKind::Pointer {
+                    *llval = bx.pointercast(*llval, ty);
+                } else {
+                    *llval = bx.bitcast(*llval, ty);
+                }
```

Comment:
Maybe it makes sense to have a more generic cast method that does the check itself?

Reply:
We basically only generate pointer casts, so I don't think such an operation would be very widely used. (All existing usage of `bitcast` that I have seen was intended to convert between pointer types, and almost all can be removed after opaque pointers, e.g. https://github.com/rust-lang/rust/pull/105412/commits/68cabc0c87fe4f89ab3a6434b5adc7d5ddf77b87)

Excluding pointer types, the only other things that can be `bitcast`ed in LLVM are differently-typed vectors `<4 x i8> to <2 x i16>` and between vectors and integers `<4 x i8> to i32`. This would probably only be used in `transmute`. (We currently codegen such transmutes as roundtrips through memory: https://godbolt.org/z/M6fr1af8z.)

I don't think we support either of those operations as part of a field extraction, so I suspect the `else` here isn't necessary. @luqmana did you encounter code that still needs the `bitcast`?

Reply:
I did not. Updated to just the pointercasts.